### PR TITLE
Add a Web UI

### LINKED ISSUE #34: Minimal log-browsing UI
I'm thinking something like the "Executors" tab in Spark. Is this in the roadmap somewhere? If yes, I could probably contribute a proof-of-concept.

### PR DESCRIPTION
**Pull request by jcrist**
This builds on #42 (thanks @superbobry), with several UX, robustness, and styling improvements. Additionally, the yarn proxy filter (`AmIpFilter`) is applied to ensure access is always authenticated (see https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/YarnApplicationSecurity.html). I have tested this filter works locally with both kerberos and simple authentication, but adding a test would require exposing the raw webui port and that seemed unnecessary.

Fixes #34.

### CODE REVIEW

**Comment by jcrist** [comment]
# Screenshots:

## Services Page

This is the initial page. Each icon represents running, succeeded, killed (should really rename this stopped), and failed containers. Tooltips indicate the meaning of each of these. I liked icons over labels as the names are rather long, and once familiar with their meanings the counts can be gotten from a quick glance.

screenshot_2018-09-11 services - skein application_1536211171559_0042

## Expanded services panel

Each service has an expandable panel. This lists two table - one of active containers (waiting, requested, and running), and one of completed containers (succeeded, killed, failed). Information and links to the running logs are provided.

screenshot_2018-09-11 services - skein application_1536211171559_0042 1

## Resized window

Because we're using bootstrap, when the window is resized the elements collapse, keeping things aesthetically nice.

screenshot_2018-09-11 services - skein application_1536211171559_0042 2

## Key/Value Store Page

The Key/Value Store page contains a table of key -> value. If a value is non-utf8 it's labeled as `<binary value>`. A client-side filter on key-name could be added later if desired.

screenshot_2018-09-11 Key/Value - skein application_1536211171559_0042

**Comment by jcrist** [comment]
I plan to leave this up for quick review, but hope to merge in the next few days. cc @superbobry.

**Comment by superbobry** [comment]
This looks awesome!

File: java/pom.xml [hunk 1]
```diff
@@ -94,6 +113,8 @@
           <compilerArgs>
             <arg>-Xlint:all</arg>
           </compilerArgs>
+          <source>7</source>
```
Comment (superbobry) [inline java/pom.xml]: Do you think we can bump this up to Java 8? 

> Oracle JDK 7 is supported across all versions of Cloudera Manager 5 and CDH 5. Oracle JDK 8 is supported in C5.3.x and higher. [[...]](https://www.cloudera.com/documentation/enterprise/release-notes/topics/rn_consolidated_pcm.html)

File: java/src/main/java/com/anaconda/skein/ApplicationMaster.java [hunk 2]
```diff
@@ -149,13 +151,62 @@ public void run() {
...
+
+    List<WebUI.ServiceContext> serviceContexts = Lists.transform(sortedServices,
+        new Function<ServiceTracker, WebUI.ServiceContext>() {
+          public WebUI.ServiceContext apply(ServiceTracker tracker) {
+            return tracker.toServiceContext();
+          }
+        }
+    );
+
+    boolean https = conf.get(
```
Comment (superbobry) [inline java/src/main/java/com/anaconda/skein/ApplicationMaster.java]: Is this variable unused?

File: java/src/main/java/com/anaconda/skein/WebUI.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    Server server = new Server(port);
+
+    // Handler for all static resources
+    ServletContextHandler context = new ServletContextHandler();
+    context.setContextPath("/");
+    // Hack to get directory containing resources, since `URI.resolve` doesn't
+    // work on opaque (e.g. non-filesystem) paths.
+    URI baseURI = URI.create(
+        WebUI.class.getResource("/META-INF/resources/favicon.ico").toURI()
+                   .toASCIIString().replaceFirst("/favicon.ico$","/")
```
Comment (superbobry) [inline java/src/main/java/com/anaconda/skein/WebUI.java]: Nitpick: whitespace after ,.

File: java/pom.xml (hunk 1 above)
Comment (jcrist) [reply java/pom.xml]: It's supported, but not mandatory. Not all deployments are necessarily running java 8 yet.

**Pull request merged**